data_structures/binary_tree/is_bst.py

```diff
@@ -1,131 +1,96 @@
-
-def is_binary_search_tree(root: TreeNode | None) -> bool:
-    """
-    >>> is_binary_search_tree(TreeNode(data=2,
-    ...                                left=TreeNode(data=1),
-    ...                                right=TreeNode(data=3))
-    ...                                )
-    True
-
-    >>> is_binary_search_tree(TreeNode(data=0,
-    ...                                left=TreeNode(data=-11),
-    ...                                right=TreeNode(data=3))
-    ...                                )
-    True
-
-    >>> is_binary_search_tree(TreeNode(data=5,
-    ...                                left=TreeNode(data=1),
-    ...                                right=TreeNode(data=4, left=TreeNode(data=3)))
-    ...                      )
-    False
+    left: Node | None = None
+    right: Node | None = None
 
-    >>> is_binary_search_tree(TreeNode(data='a',
-    ...                                left=TreeNode(data=1),
-    ...                                right=TreeNode(data=4, left=TreeNode(data=3)))
-    ...                      )
-    Traceback (most recent call last):
-     ...
-    ValueError: Each node should be type of TreeNode and data should be float.
-
-    >>> is_binary_search_tree(TreeNode(data=2,
-    ...                                left=TreeNode([]),
-    ...                                right=TreeNode(data=4, left=TreeNode(data=3)))
-    ...                                )
-    Traceback (most recent call last):
-     ...
-    ValueError: Each node should be type of TreeNode and data should be float.
-    """
-
-    # Validation
-    def is_valid_tree(node: TreeNode | None) -> bool:
+    def __iter__(self) -> Iterator[float]:
         """
-        >>> is_valid_tree(None)
-        True
-        >>> is_valid_tree('abc')
-        False
-        >>> is_valid_tree(TreeNode(data='not a float'))
-        False
-        >>> is_valid_tree(TreeNode(data=1, left=TreeNode('123')))
-        False
+        >>> root = Node(data=2.1)
+        >>> list(root)
+        [2.1]
+        >>> root.left=Node(data=2.0)
+        >>> list(root)
+        [2.0, 2.1]
+        >>> root.right=Node(data=2.2)
+        >>> list(root)
+        [2.0, 2.1, 2.2]
         """
-        if node is None:
-            return True
-
-        if not isinstance(node, TreeNode):
-            return False
-
-        try:
-            float(node.data)
-        except (TypeError, ValueError):
-            return False
-
-        return is_valid_tree(node.left) and is_valid_tree(node.right)
-
-    if not is_valid_tree(root):
-        raise ValueError(
-            "Each node should be type of TreeNode and data should be float."
-        )
-
-    def is_binary_search_tree_recursive_check(
-        node: TreeNode | None, left_bound: float, right_bound: float
-    ) -> bool:
+        if self.left:
+            yield from self.left
+        yield self.data
+        if self.right:
+            yield from self.right
+
+    @property
+    def is_sorted(self) -> bool:
```

Comment:
The name of this function no longer matches the name of the file, and that might confuse some readers. I feel that either this function should be renamed to `is_bst` or the file should be renamed to `is_sorted_binary_tree`.

Reply:
I dislike using an acronym and `binary_tree.is_sorted_binary_tree` is too long and awkward for a property of `BinaryTree` so let's go for a filename change.